packages/@ember/object/index.ts

```diff
@@ -130,18 +131,43 @@ export {
 
 const BINDINGS_MAP = new WeakMap();
 
-function setupAction(target, key, actionFn) {
-  if (target.constructor !== undefined && typeof target.constructor.proto === 'function') {
+interface HasProto {
+  constructor: {
+    proto(): void;
+  };
+}
+
+function hasProto(obj: unknown): obj is HasProto {
+  return (
+    obj != null &&
+    (obj as any).constructor !== undefined &&
+    typeof ((obj as any).constructor as any).proto === 'function'
+  );
+}
+
+interface HasActions {
+  actions: Record<string | symbol, unknown>;
+}
+
+function setupAction(
+  target: object,
+  key: string | symbol,
+  actionFn: Function
+): TypedPropertyDescriptor<unknown> {
+  // SAFETY: Accessing an unknown property on a non-null value isn't unsafe.
+  if (hasProto(target)) {
     target.constructor.proto();
   }
 
+  let targetWithActions = target as HasActions;
```

Comment:
By contrast, should there be a guard here? Even just `assert()`-ing that `'actions' in target && typeof (target as HasActions).actions === 'object'` or something?

Reply:
Well not exactly, since we immediately set actions if it is not set, but I admit that this is kind of awkward.